server/somesockets.py

```diff
@@ -1,47 +1,87 @@
-from flask import Flask, render_template
+from flask import Flask, render_template, request
 from flask_socketio import SocketIO, send, emit
+from secrets import SOCKETIO_SECRET_KEY
+import models # import like this or you'll get circular dependencies
+from flask_mongoengine import MongoEngine
+from urllib.parse import urlparse, parse_qs
 
+
+#=====================================================
+# Global vars
+#=====================================================
 app = Flask(__name__)
-app.config['SECRET_KEY'] = 'secret!'
+app.config['SECRET_KEY'] = SOCKETIO_SECRET_KEY
+app.config.from_object('config')
+
+db = MongoEngine(app)
+
 socketio = SocketIO(app)
 
+
 #=====================================================
-# SocketIO
+# Error handling
 #=====================================================
+@socketio.on_error()
+def error_handler(e):
+    if type(e) in [ValueError, TypeError]:
+        emit('error', 'Malformed request')
+    else:
+        print(e)
+        emit('error', 'General error, try again')
 
-socketio = SocketIO(app)
 
+@socketio.on_error_default
+def default_error_handler(e):
+    app.logger.error("General error with request.sid=" + request.sid)
+    emit('error', 'Internal server error, try again')
+
+
+#=====================================================
+# Routes
+#=====================================================
 @socketio.on('connect')
-def handle_connect():
-    print("a user connected")
+def connect():
+    code = request.query_string.decode().lower()
+    code = parse_qs(urlparse(code).query)
+    if code.get('code') is None:
+        error_handler(ValueError('Invalid room code'))
+        return
+
+    code = models.Session.objects(code=code['code']).first()
+    if code is None:
+        emit('roomNotFound', 'No session exists for this room code')
+        return
+
+    join_room(code)
+    emit('connect', dict(message="Another user connected"), room=code)
 
 @socketio.on('disconnect')
-def handle_disconnect():
-    print("a user has left")
+def disconnect():
+    # This function technically doesn't even need to exist;
+    # by default socketio handles room disconnection.
+    # This is here purely in case we want to do anything additional.
+    pass
 
 @socketio.on('update')
-def handle_update(json):
-    print('received json: ' + str(json))
-    emit('update', json)
-
-@socketio.on('FindRoom')
-def handle_FindRoom(json):
-    print('received room: ' + str(json))
-    
-#TODO: Send this in the event the submitted room code isn't found
-def handle_RoomNotFound(json):
-    emit('RoomNotFound', json)
-    print ('Room Not Found!')
+def update(map_id, room):
+    """
+    This isn't a socket event, this is triggered from within the flask application.
+    Nobody should be hitting this endpoint.
+    """
+    map = models.GameMap.objects(id=map_id).first()
+    emit('update', json=map, room=room)
```

Comment:
im an idiot i forgot to wrap the map_id in a dict. that might be one of a few issues now that I realize it

Reply:
@joshherkness i will make some updates and then you should try to see if they make a difference within the next hour or so